app/process_proposal.go

```diff
@@ -81,11 +81,18 @@ func (app *App) ProcessProposal(req abci.RequestProcessProposal) abci.ResponsePr
 
 	// iterate over all of the MsgPayForBlob transactions and ensure that their
 	// commitments are subtree roots of the data root.
+	seenFirstPFB := false
 	for _, rawTx := range req.BlockData.Txs {
 		tx := rawTx
 		wrappedTx, isWrapped := coretypes.UnmarshalIndexWrapper(rawTx)
 		if isWrapped {
+			seenFirstPFB = true
 			tx = wrappedTx.Tx
+		} else if seenFirstPFB {
+			logInvalidPropBlock(app.Logger(), req.Header, "Wrapped PFBs must come at the end of the block")
+			return abci.ResponseProcessProposal{
+				Result: abci.ResponseProcessProposal_REJECT,
+			}
```

Comment:
[no change needed][optional] This works as-is but it adds complexity to a `for` loop that is performing a different check. Since the check for PFB tx ordering seems distinct, it may be extracted like:


```go
	if !arePFBsOrderedAfterTxs(req.BlockData.Txs) {
		logInvalidPropBlock(app.Logger(), req.Header, "Wrapped PFBs must come at the end of the block")
		return abci.ResponseProcessProposal{
			Result: abci.ResponseProcessProposal_REJECT,
		}
	}
```


```go
func arePFBsOrderedAfterTxs(txs [][]byte) bool {
	seenFirstPFB := false
	for _, rawTx := range txs {
		_, isWrapped := coretypes.UnmarshalIndexWrapper(rawTx)
		if isWrapped {
			seenFirstPFB = true
		} else if seenFirstPFB {
			return false
		}
	}
	return true
}

```

Reply:
Yeah I actually wrote something similar to this at first. The reason why I piggybacked off the for loop is because we're already identifying whether a tx is wrapped or not and I wanted to avoid needlessly unmarshalling the same proto bytes multiple times. 

If I recall correctly, we're already using that for loop for multiple validity checks

Reply:
Agreed, the proposal may be a performance penalty. Without measuring the impact it seems preferable to opt for code clarity over potential performance hit. If needed in the future, we can benchmark and optimize for performance. At that time, it will be easier to optimize the clearer code.

Reply:
apologies, I read the diff wrong in one of the test files, I thought we had it and removed it but that's not the case.

Reply:
Sure I don't mind doing that.